ibis/common/grounds.py

```diff
@@ -167,10 +175,23 @@ def __reduce__(self):
         kwargs = dict(zip(self.argnames, self.args))
         return (self._reconstruct, (kwargs,))
 
+    # TODO(kszucs): consider to make a separate mixin class for this
+    def copy(self, **overrides):
+        kwargs = dict(zip(self.argnames, self.args))
+        newargs = {**kwargs, **overrides}
+        return self.__class__(**newargs)
+
 
-class Singleton(Base):
+class Weakrefable(Base):
 
     __slots__ = ('__weakref__',)
+
+
+class Singleton(Weakrefable):
+    # NOTE: this only considers the input arguments, when combined with
+    # Annotable base class Singleton must come after in the MRO
```

Comment:
This comment is a tad hard to parse. Is it saying that the following will fail?

```python
class Foo(Annotable, Singleton):
    ...
```

Reply:
Yes, so that `Singleton.__create__` gets called before `Annotable.__create__`, otherwise the argument validation cannot be spared.